File: packages/ast/src/canonical.ts
/**
 * AST canonicalization and content hashing.
 *
 * This hash is both the result-cache key and drill-down token. **The Python compiler must
 * produce the same value for the same AST.** Conformance tests enforce these rules.
 *
 * Canonical-form rules:
 *   R1. Remove `span`.
 *   R2. Remove `raw` by default so equivalent duration spellings share a cache entry.
 *   R3. Treat `undefined` and absent keys equally.
 *   R4. Sort object keys lexicographically.
 *   R5. Serialize integral floating-point values as integers.
 *   R6. Preserve array order because `A AND B` and `B AND A` retain source ordering.
 */
import type { Node } from './nodes.js';
import { sha256Hex } from './sha256.js';

export interface CanonicalizeOptions {
  /** Whether to remove original spelling (`raw`); hashes remove it and round-trip checks retain it. */
  readonly dropRaw?: boolean;
}

/** Keys always removed from canonical form. */
const ALWAYS_DROP = new Set(['span']);
/** Additional keys removed when `dropRaw` is enabled. */
const PRESENTATION_KEYS = new Set(['raw']);

export function canonicalize(value: unknown, options: CanonicalizeOptions = {}): unknown {
  const dropRaw = options.dropRaw ?? true;
  return normalize(value, dropRaw);
}

function normalize(value: unknown, dropRaw: boolean): unknown {
  if (value === null) return null;
  if (typeof value === 'number') return normalizeNumber(value);
  if (typeof value !== 'object') return value;
  if (Array.isArray(value)) return value.map((v) => normalize(v, dropRaw));

  const out: Record<string, unknown> = {};
  for (const key of Object.keys(value as Record<string, unknown>).sort()) {
    if (ALWAYS_DROP.has(key)) continue;
    if (dropRaw && PRESENTATION_KEYS.has(key)) continue;
    const v = (value as Record<string, unknown>)[key];
    if (v === undefined) continue;
    out[key] = normalize(v, dropRaw);
  }
  return out;
}

/**
 * R5. Fold `1.0` into `1`.
 *
 * JavaScript serializes `1.0` as `"1"`, while Python normally emits `"1.0"`.
 * Folding prevents cross-language hash differences.
 */
function normalizeNumber(value: number): number {
  if (!Number.isFinite(value)) throw new Error(`Cannot normalize a non-finite number: ${value}`);
  if (value !== 0 && (Math.abs(value) < 1e-6 || Math.abs(value) >= 1e21)) {
    throw new Error(`Cannot normalize a number outside the canonical range: ${value}`);
  }
  return Number.isInteger(value) ? Math.trunc(value) : value;
}

/** Key-sorted serialization of an already normalized value. */
export function canonicalStringify(value: unknown): string {
  if (value === null) return 'null';
  if (typeof value === 'number') return numberToJson(value);
  if (typeof value === 'string') return JSON.stringify(value);
  if (typeof value === 'boolean') return value ? 'true' : 'false';
  if (Array.isArray(value)) return '[' + value.map(canonicalStringify).join(',') + ']';
  if (typeof value === 'object') {
    const entries = Object.entries(value as Record<string, unknown>)
      .filter(([, v]) => v !== undefined)
      .sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0));
    return (
      '{' + entries.map(([k, v]) => JSON.stringify(k) + ':' + canonicalStringify(v)).join(',') + '}'
    );
  }
  throw new Error(`Cannot serialize value: ${String(value)}`);
}

function numberToJson(value: number): string {
  // Normalization has already folded integral values. JavaScript and Python both use a
  // shortest round-trippable representation for the remaining floats.
  return Object.is(value, -0) ? '0' : String(value);
}

/** Canonical JSON string; Python must produce the identical string. */
export function canonicalJson(node: Node, options?: CanonicalizeOptions): string {
  return canonicalStringify(canonicalize(node, options));
}

/** Content hash used by cache keys and drill-down tokens. */
export function canonicalHash(node: Node): string {
  return 'sha256:' + sha256Hex(canonicalStringify(normalizeCommutativeBoolean(canonicalize(node))));
}

/** Sorts only commutative boolean operands for cache hashing; presentation order stays intact. */
function normalizeCommutativeBoolean(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(normalizeCommutativeBoolean);
  if (value === null || typeof value !== 'object') return value;

  const object = Object.fromEntries(
    Object.entries(value as Record<string, unknown>).map(([key, child]) => [
      key,
      normalizeCommutativeBoolean(child),
    ]),
  );
  if (
    object.kind !== 'BinaryExpr' ||
    (object.op !== 'AND' && object.op !== 'OR') ||
    object.left === undefined ||
    object.right === undefined
  ) {
    return object;
  }

  const op = object.op;
  const flatten = (node: unknown): unknown[] => {
    if (
      node !== null &&
      typeof node === 'object' &&
      !Array.isArray(node) &&
      (node as Record<string, unknown>).kind === 'BinaryExpr' &&
      (node as Record<string, unknown>).op === op
    ) {
      const binary = node as Record<string, unknown>;
      return [...flatten(binary.left), ...flatten(binary.right)];
    }
    return [node];
  };
  const operands = flatten(object).sort((a, b) => {
    const left = canonicalStringify(a);
    const right = canonicalStringify(b);
    return left < right ? -1 : left > right ? 1 : 0;
  });
  let result = operands[0]!;
  for (let index = 1; index < operands.length; index++) {
    result = { kind: 'BinaryExpr', op, left: result, right: operands[index]! };
  }
  return result;
}

/**
 * Semantic equality ignoring spans and original spelling. Used to suppress no-op commits
 * in bidirectional builder/DSL synchronization.
 */
export function astEquals(a: Node, b: Node): boolean {
  return canonicalJson(a) === canonicalJson(b);
}

/**
 * Structural equality including original spelling. This stricter comparison is used by
 * round-trip property P1 to detect printer spelling changes.
 */
export function astStrictEquals(a: Node, b: Node): boolean {
  return canonicalJson(a, { dropRaw: false }) === canonicalJson(b, { dropRaw: false });
}

/**
 * Deterministic binding ID.
 *
 * Random IDs break print/parse equality and card identity. Repeated references to the same
 * event must share an ID so the compiler can reuse one witness CTE.
 */
export function makeBindingId(parts: {
  readonly scope: string | null;
  readonly eventType: string;
  readonly ordinal: string | number;
}): string {
  const scope = parts.scope ?? 'any';
  const ordinal = typeof parts.ordinal === 'number' ? `nth-${parts.ordinal}` : parts.ordinal;
  return `${scope}.${parts.eventType}#${ordinal}`;
}

File: services/api/src/lod_api/canonical.py
"""Python counterpart of TypeScript AST canonicalization and hashing.

The hash keys result caches and drill-down tokens, so both languages must produce identical output.
Rules remove spans and presentation spelling, sort object keys, fold integral floats, and preserve
array order.
"""

from __future__ import annotations

import hashlib
import math
from decimal import Decimal
from typing import Any

ALWAYS_DROP = frozenset({"span"})
PRESENTATION_KEYS = frozenset({"raw"})
MIN_CANONICAL_ABS = 1e-6
MAX_CANONICAL_ABS = 1e21


def canonicalize(value: Any, *, drop_raw: bool = True) -> Any:
    """Convert a value to a JSON-serializable canonical form."""
    if value is None:
        return None
    if isinstance(value, bool):
        return value
    if isinstance(value, int):
        if abs(value) >= MAX_CANONICAL_ABS:
            raise ValueError(
                f"Canonical integers must have an absolute value below {MAX_CANONICAL_ABS}: {value}"
            )
        return value
    if isinstance(value, float):
        return _normalize_number(value)
    if isinstance(value, str):
        return value
    if isinstance(value, (list, tuple)):
        return [canonicalize(v, drop_raw=drop_raw) for v in value]
    if isinstance(value, dict):
        out: dict[str, Any] = {}
        for key in sorted(value):
            if key in ALWAYS_DROP:
                continue
            if drop_raw and key in PRESENTATION_KEYS:
                continue
            v = value[key]
            if v is None and key not in _NULLABLE_KEYS:
                # TypeScript removes undefined but retains meaningful null AST fields.
                out[key] = None
                continue
            out[key] = canonicalize(v, drop_raw=drop_raw)
        return out
    raise TypeError(f"Cannot canonicalize value of type {type(value).__name__}")


#: Keys where null is meaningful; all nulls are currently retained.
_NULLABLE_KEYS: frozenset[str] = frozenset()


def _normalize_number(value: float) -> int | float:
    """Fold integral floats and reject values whose JS/Python spellings diverge."""
    if not math.isfinite(value):
        raise ValueError(f"Cannot canonicalize non-finite number: {value}")
    magnitude = abs(value)
    if magnitude != 0 and not MIN_CANONICAL_ABS <= magnitude < MAX_CANONICAL_ABS:
        raise ValueError(
            "Canonical numbers must be zero or have an absolute value in "
            f"[{MIN_CANONICAL_ABS}, {MAX_CANONICAL_ABS}): {value}"
        )
    if value.is_integer():
        return int(value)
    return value


def canonical_stringify(value: Any) -> str:
    """Serialize sorted keys identically to TypeScript `canonicalStringify`."""
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, int):
        return str(value)
    if isinstance(value, float):
        return _float_to_json(value)
    if isinstance(value, str):
        return _string_to_json(value)
    if isinstance(value, list):
        return "[" + ",".join(canonical_stringify(v) for v in value) + "]"
    if isinstance(value, dict):
        parts = [
            _string_to_json(k) + ":" + canonical_stringify(v) for k, v in sorted(value.items())
        ]
        return "{" + ",".join(parts) + "}"
    raise TypeError(f"Cannot serialize value of type {type(value).__name__}")


def _float_to_json(value: float) -> str:
    """Match JavaScript's shortest round-trippable float representation."""
    if value == 0:
        return "0"
    text = repr(value)
    # ECMAScript uses fixed notation for the allowed range even when Python's repr chooses an
    # exponent. Expanding the shortest Python representation preserves its exact decimal digits.
    if "e" in text or "E" in text:
        return format(Decimal(text), "f")
    return text


#: Escapes used by JSON.stringify, implemented directly because Python differs subtly.
_ESCAPES = {
    '"': '\\"',
    "\\": "\\\\",
    "\n": "\\n",
    "\r": "\\r",
    "\t": "\\t",
    "\b": "\\b",
    "\f": "\\f",
}


def _string_to_json(value: str) -> str:
    out = ['"']
    for ch in value:
        escaped = _ESCAPES.get(ch)
        if escaped is not None:
            out.append(escaped)
        elif ord(ch) < 0x20:
            out.append(f"\\u{ord(ch):04x}")
        else:
            # JSON.stringify preserves non-ASCII text such as localized labels.
            out.append(ch)
    out.append('"')
    return "".join(out)


def canonical_json(ast: Any, *, drop_raw: bool = True) -> str:
    return canonical_stringify(canonicalize(ast, drop_raw=drop_raw))


def canonical_hash(ast: Any) -> str:
    """Return the content hash used by cache keys and drill-down tokens."""
    digest = hashlib.sha256(
        canonical_json(_normalize_commutative_for_hash(ast)).encode("utf-8")
    ).hexdigest()
    return f"sha256:{digest}"


def _normalize_commutative_for_hash(value: Any) -> Any:
    """Sort associative boolean operands for semantic cache-key equivalence."""
    if isinstance(value, list):
        return [_normalize_commutative_for_hash(item) for item in value]
    if not isinstance(value, dict):
        return value

    normalized = {key: _normalize_commutative_for_hash(item) for key, item in value.items()}
    op = normalized.get("op")
    if normalized.get("kind") != "BinaryExpr" or op not in {"AND", "OR"}:
        return normalized

    operands: list[Any] = []

    def collect(node: Any) -> None:
        if isinstance(node, dict) and node.get("kind") == "BinaryExpr" and node.get("op") == op:
            collect(node.get("left"))
            collect(node.get("right"))
        else:
            operands.append(node)

    collect(normalized)
    operands.sort(key=lambda item: canonical_stringify(canonicalize(item)))
    result = operands[0]
    for operand in operands[1:]:
        result = {"kind": "BinaryExpr", "op": op, "left": result, "right": operand}
    return result
